contracts/Marketplace.sol

```diff
@@ -416,4 +421,66 @@ contract Marketplace is Ownable, IMarketplace {
     function reInitialize(address datacoinAddress, address currencyUpdateAgentAddress, address prev_marketplace_address) public onlyOwner {
         _initialize(datacoinAddress, currencyUpdateAgentAddress, prev_marketplace_address);
     }
+
+    //whitelist functionality
+
+    //should the whitelist getter be built into getProduct()?
+    function getRequiresWhitelist(bytes32 productId) public view returns (bool) {
+        (,address _owner,,,,,) = getProduct(productId);
+        require(_owner != 0x0, "error_notFound");
+        //if it's not local this will return 0, which is false
+        Product storage p = products[productId];
+        return p.requiresWhitelist;
+    }
+
+    function setRequiresWhitelist(bytes32 productId, bool _requiresWhitelist) public onlyProductOwner(productId) {
+        _importProductIfNeeded(productId);
+        Product storage p = products[productId];
+        require(p.id != 0x0, "Product not found");
```

Comment:
please use "error codes" and if you add new error codes, please add them to `strings/Marketplace.json`. This was originally a recommendation by the Marketplace contract auditor, and while at that time I was a bit meh about it, I think it's overall a good practice not to "localize by default". Error codes should begin with error_ and otherwise be camel case. I see you've already done that with BancorAdaptor.

Reply:
one reason could also to make it easier to keep them uniform, for automatic matching later. For instance, "Product not found!" could be different from "Product not found." not for a human but for a machine. Those error codes are in that sense targeted to machines. Not sure if that's 100% true, but error codes should be added to tests, so in that sense at least yes.